### PR TITLE
[runtime] Fix aot compiler leaks

### PR DESCRIPTION
**Pull request by alexanderkyte**
These are the most prominent leaks in the AOT compiler, and represent a fairly significant memory footprint.

### CODE REVIEW

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -7375,6 +7376,39 @@ add_gsharedvt_wrappers (MonoAotCompile *acfg, MonoMethodSignature *sig, gboolean
 	}
 }
 
+static void mono_destroy_compile_aot (MonoAotCompile *acfg)
+{
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: This should have a newline after void.
There is no need to prefix static functions with 'mono_'.

File: mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -7375,6 +7376,39 @@ add_gsharedvt_wrappers (MonoAotCompile *acfg, MonoMethodSignature *sig, gboolean
...
+	GSList *curr = acfg->destroy_queue;
+
+	while (curr != NULL) {
+		mono_destroy_compile (curr->data);
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: It would look better to iterate over the list, then delete it using g_slist_free ().

File: mono/mini/aot-compiler.c [hunk 3]
```diff
@@ -7375,6 +7376,39 @@ add_gsharedvt_wrappers (MonoAotCompile *acfg, MonoMethodSignature *sig, gboolean
...
+	gboolean done = FALSE;
+
+	while (!done) {
+		gpointer oldHead = acfg->destroy_queue;
+		GSList *head = g_new0 (GSList, 1);
+		head->data = cfg;
+		head->next = (GSList *)oldHead;
+
+		gpointer prev = InterlockedCompareExchangePointer(&acfg->destroy_queue, head, oldHead);
+
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: This could use mono_acfg_lock () instead of lockless code.

File: mono/mini/aot-compiler.c [hunk 4]
```diff
@@ -7509,7 +7544,7 @@ compile_method (MonoAotCompile *acfg, MonoMethod *method)
 		if (acfg->aot_opts.print_skipped_methods)
 			printf ("Skip (abs call): %s\n", mono_method_get_full_name (method));
 		InterlockedIncrement (&acfg->stats.abscount);
-		mono_destroy_compile (cfg);
+		pending_compile_destroy (acfg, cfg);
 		return;
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: Why is this needed ? Keeping cfg-s around will dramatically increase the memory usage of the aot compiler.

File: mono/mini/mini.c [hunk 5]
```diff
@@ -3365,6 +3365,7 @@ mini_method_compile (MonoMethod *method, guint32 opts, MonoDomain *domain, JitFl
 				if (cfg->verbose_level >= (cfg->llvm_only ? 0 : 1)) {
 					//nm = mono_method_full_name (cfg->method, TRUE);
 					printf ("LLVM failed for '%s': %s\n", method->name, cfg->exception_message);
+					g_free (cfg->exception_message);
 					//g_free (nm);
```
Comment (vargaz) [inline mono/mini/mini.c]: This should already be freed by mono_destroy_compile ().

File: mono/mini/mini-llvm.c [hunk 6]
```diff
@@ -6600,6 +6600,14 @@ free_ctx (EmitContext *ctx)
 	g_hash_table_destroy (ctx->region_to_handler);
 	g_hash_table_destroy (ctx->clause_to_handler);
 	g_hash_table_destroy (ctx->jit_callees);
+
+	GHashTableIter iter;
+	gpointer key;
+	g_hash_table_iter_init (&iter, ctx->method_to_callers);
+	while (g_hash_table_iter_next (&iter, &key, &l))
+		g_slist_free (l);
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: You can pass NULL instead of &key if its not needed.

File: mono/mini/aot-compiler.c (hunk 4 above)
Comment (vargaz) [inline mono/mini/aot-compiler.c]: So these are the failed compiles right ?

**Comment by vargaz** [comment]
This could use a rebase, since many changes change previous ones.

File: mono/mini/aot-compiler.c (hunk 4 above)
Comment (alexanderkyte) [inline mono/mini/aot-compiler.c]: These are failed compiles, but the types that these refer to are freed by mono_destroy_compile. If we free them earlier, we end up with memory unsafety. Currently we don't free them later I'm pretty sure, we leak them. I'm going to figure out why we're not reparenting the types though, we should be able to promptly free the memory.

File: mono/mini/mini.c (hunk 5 above)
Comment (alexanderkyte) [inline mono/mini/mini.c]: Instruments informed me that it was leaked.

Comment (vargaz) [inline mono/mini/mini.c]: Maybe its only leaked in the llvm_only case, which doesn't free the cfg.

**Comment by vargaz** [comment]
Looks good.
